cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```
@@ -0,0 +1,497 @@
+  }
+
+  @Inject(optional = true)
+  public void setHostname(@Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress host) {
+    if (Objects.nonNull(host)) {
+      this.hostname = host.getCanonicalHostName();
+    }
+  }
+
+  @Override
+  public ProgramController dispatchProgram(ProgramRunDispatcherInfo programRunDispatcherInfo) throws Exception {
+    RunId runId = programRunDispatcherInfo.getRunId();
+    LOG.debug("Preparing to dispatch program run: {}", runId);
+    ProgramDescriptor programDescriptor = programRunDispatcherInfo.getProgramDescriptor();
+    ProgramOptions options = programRunDispatcherInfo.getProgramOptions();
+    boolean isDistributed = programRunDispatcherInfo.isDistributed();
+    ProgramId programId = programDescriptor.getProgramId();
+    ClusterMode clusterMode = ProgramRunners.getClusterMode(options);
+    ProgramRunnerFactory progRunnerFactory = (clusterMode == ClusterMode.ON_PREMISE ? programRunnerFactory
+      : Optional.ofNullable(remoteProgramRunnerFactory).orElseThrow(UnsupportedOperationException::new));
+
+    String peer = options.getArguments().getOption(ProgramOptionConstants.PEER_NAME);
+    if (peer != null) {
+      try {
+        // For tethered pipeline runs, fetch artifacts from ArtifactCacheService
+        String basePath = String.format("%s/peers/%s", Constants.Gateway.INTERNAL_API_VERSION_3, peer);
+        RemoteClient remoteClient = remoteClientFactory.createRemoteClient(Constants.Service.ARTIFACT_CACHE_SERVICE,
+          RemoteClientFactory.NO_VERIFY_HTTP_REQUEST_CONFIG, basePath);
+        RemoteArtifactRepositoryReader artifactRepositoryReader = new RemoteArtifactRepositoryReader(
+          locationFactory, remoteClient);
+        artifactRepository = new RemoteArtifactRepository(cConf, artifactRepositoryReader,
+          progRunnerFactory);
+      } catch (Exception e) {
+        LOG.error("Exception while trying to create remote client factory", e);
```

log and rethrow is a bad pattern, it results in the error being logged multiple times in multiple places. It's better to let it propagate up and make the caller handle it.

---

Changed.